src/finite_field.rs

```diff
@@ -11,24 +11,29 @@ pub enum FiniteFieldError {
     InputSizeMismatch,
 }
 
-/// Newtype wrapper over u32
-///
-/// Implements the arithmetic over the finite prime field
-#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
-pub struct Field(u32);
+use crate::fp::Fp;
+
+const SMALL_FP: Fp = Fp {
+    p: 4293918721,
+    mu: 17302828673139736575,
+    r2: 1676699750,
+};
 
 /// Modulus for the field, a FFT friendly prime: 2^32 - 2^20 + 1
-pub const MODULUS: u32 = 4293918721;
+pub const MODULUS: u32 = SMALL_FP.p as u32;
 /// Generator for the multiplicative subgroup
 pub(crate) const GENERATOR: u32 = 3925978153;
 /// Number of primitive roots
 pub(crate) const N_ROOTS: u32 = 1 << 20; // number of primitive roots
 
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, Hash, Default)]
+pub struct Field(u128);
+
 impl std::ops::Add for Field {
-    type Output = Field;
+    type Output = Self;
 
     fn add(self, rhs: Self) -> Self {
-        self - Field(MODULUS - rhs.0)
+        Self(SMALL_FP.add(self.0, rhs.0))
```

Comment:
This means that all instances of Field will use SMALL_FP modulus, right? so how to express other instances with different modulus, but same underlying implementation. Adding some examples can help to understand its usage.

Reply:
This is the bit I'm struggling with :)

What I want is to use generics, somehow, to replace `SMALL_FP` with any old instance of `Fp`.